pkg/provisioner/ironic/ironic.go

```diff
@@ -508,61 +500,61 @@ func (p *ironicProvisioner) ValidateManagementAccess(credentialsChanged, force b
 		provID = ironicNode.UUID
 
 		if ironicNode.Name != p.ironicNodeNameFromHost() {
-			updates := nodes.UpdateOpts{
+			updates = append(
+				updates,
 				nodes.UpdateOperation{
 					Op:    nodes.ReplaceOp,
 					Path:  "/name",
 					Value: p.ironicNodeNameFromHost(),
 				},
-			}
-			ironicNode, err = nodes.Update(p.client, ironicNode.UUID, updates).Extract()
-			switch err.(type) {
-			case nil:
-			case gophercloud.ErrDefault409:
-				p.log.Info("could not update ironic node name, busy")
-				result, err = retryAfterDelay(provisionRequeueDelay)
-				return
-			default:
-				result, err = transientError(errors.Wrap(err, "failed to update ironc node name"))
-				return
-			}
-			p.log.Info("updated ironic node name")
-
+			)
 		}
 
 		// Look for the case where we previously enrolled this node
 		// and now the credentials have changed.
 		if credentialsChanged {
-			updates := nodes.UpdateOpts{
+			updates = append(
+				updates,
 				nodes.UpdateOperation{
 					Op:    nodes.ReplaceOp,
 					Path:  "/driver_info",
 					Value: driverInfo,
 				},
-			}
-			ironicNode, err = nodes.Update(p.client, ironicNode.UUID, updates).Extract()
-			switch err.(type) {
-			case nil:
-			case gophercloud.ErrDefault409:
-				p.log.Info("could not update host driver settings, busy")
-				result, err = retryAfterDelay(provisionRequeueDelay)
-				return
-			default:
-				result, err = transientError(errors.Wrap(err, "failed to update host driver settings"))
-				return
-			}
-			p.log.Info("updated host driver settings")
+			)
 			// We don't return here because we also have to set the
 			// target provision state to manageable, which happens
 			// below.
 		}
 	}
+	p.log.Info("setting automated cleaning mode to",
+		"ID", ironicNode.UUID,
+		"mode", p.host.Spec.AutomatedCleaningMode)
+
+	value := p.host.Spec.AutomatedCleaningMode != metal3v1alpha1.CleaningModeDisabled
+	updates = append(
+		updates,
+		nodes.UpdateOperation{
+			Op:    nodes.ReplaceOp,
+			Path:  "/automated_clean",
+			Value: value,
+		},
```

Comment:
This UpdateOperation is still applied unconditionally, should we be checking `ironicNode.AutomatedClean` and only setting it while needed?

Reply:
Yeah, you also need to check for a non-nil pointer before dereferencing it.

Reply:
> Yeah, you also need to check for a non-nil pointer before dereferencing it.

We will never be able to do the update if we check that, because `automated_clean` = `nil` by default in gophercloud.

Reply:
If it's nil you always update:

```
if ironicNode.AutomatedClean == nil || (p.host.Spec.AutomatedCleaningMode == "disabled" && *ironicNode.AutomatedClean != false) || (p.host.Spec.AutomatedCleaningMode == "enabled" && *ironicNode.AutomatedClean == false) {
```

Reply:
🤦‍♂️, thank you!